File: trustpoint/workflows/services/validators/step_types/logic.py
"""Validation for Logic step params."""

from __future__ import annotations

from typing import Any

from django.utils.translation import gettext as _

from workflows.services.validators.common import error, is_bare_var_path

_ALLOWED_OPS: set[str] = {
    'eq',
    'ne',
    'lt',
    'lte',
    'gt',
    'gte',
    'and',
    'or',
    'not',
    'exists',
    'truthy',
    'falsy',
}


def validate_logic_step(
    *,
    idx: int,
    step_id: str,
    params: dict[str, Any],
    errors: list[str],
    steps: list[Any],
) -> None:
    """Validate Logic step parameters (Pass-1).

    Supports:
    - rules: list of rule objects
    - default: object (required)
    - rule.when: either an expr object/primitive OR a non-empty list of exprs (AND sugar)
    - actions: list (supports set.assign only)
    - then: pass/goto/stop (exactly one)
    - goto must be forward-only and refer to an existing step id
    """
    default = params.get('default')
    if not isinstance(default, dict):
        error(errors, _('Step #%s (Logic): default is required and must be an object.') % idx)

    rules = params.get('rules', [])
    if rules is None:
        rules = []
    if not isinstance(rules, list):
        error(errors, _('Step #%s (Logic): rules must be an array if provided.') % idx)
        return

    step_ids: set[str] = {
        str(s.get('id'))
        for s in steps
        if isinstance(s, dict) and isinstance(s.get('id'), str) and str(s.get('id')).strip()
    }
    index_map: dict[str, int] = {}
    for i, s in enumerate(steps):
        if isinstance(s, dict) and isinstance(s.get('id'), str):
            sid = str(s.get('id'))
            if sid.strip():
                index_map[sid] = i

    # Validate rules
    for j, rule in enumerate(rules, start=1):
        if not isinstance(rule, dict):
            error(errors, _('Step #%s (Logic): rule #%s must be an object.') % (idx, j))
            continue

        when = rule.get('when')
        if when is None:
            error(errors, _('Step #%s (Logic): rule #%s is missing "when".') % (idx, j))
        else:
            _validate_when(idx=idx, label=f'rule #{j} when', when=when, errors=errors)

        actions = rule.get('actions', [])
        _validate_actions(idx=idx, label=f'rule #{j}', actions=actions, errors=errors)

        then = rule.get('then')
        _validate_then(
            idx=idx,
            label=f'rule #{j}',
            then=then,
            errors=errors,
            step_id=step_id,
            step_ids=step_ids,
            index_map=index_map,
        )

    # Validate default
    if isinstance(default, dict):
        actions = default.get('actions', [])
        _validate_actions(idx=idx, label='default', actions=actions, errors=errors)

        then = default.get('then')
        _validate_then(
            idx=idx,
            label='default',
            then=then,
            errors=errors,
            step_id=step_id,
            step_ids=step_ids,
            index_map=index_map,
        )


def _validate_when(*, idx: int, label: str, when: Any, errors: list[str]) -> None:
    """Validate `when` which may be an expr or a list-of-exprs (AND sugar)."""
    if isinstance(when, list):
        if not when:
            error(errors, _('Step #%s (Logic): %s must not be an empty array.') % (idx, label))
            return
        for i, e in enumerate(when, start=1):
            _validate_expr(idx=idx, label=f'{label}[{i}]', expr=e, errors=errors)
        return

    _validate_expr(idx=idx, label=label, expr=when, errors=errors)


def _validate_actions(*, idx: int, label: str, actions: Any, errors: list[str]) -> None:
    if actions is None:
        return
    if not isinstance(actions, list):
        error(errors, _('Step #%s (Logic): %s actions must be an array.') % (idx, label))
        return

    for k, act in enumerate(actions, start=1):
        if not isinstance(act, dict):
            error(errors, _('Step #%s (Logic): %s action #%s must be an object.') % (idx, label, k))
            continue

        t = str(act.get('type') or '').strip().lower()
        if t != 'set':
            error(errors, _('Step #%s (Logic): %s action #%s has unknown type "%s".') % (idx, label, k, t))
            continue

        assign = act.get('assign')
        if not isinstance(assign, dict):
            error(errors, _('Step #%s (Logic): %s action #%s set.assign must be an object.') % (idx, label, k))
            continue

        for raw_key, raw_val in assign.items():
            key = str(raw_key).strip()
            if not key or not is_bare_var_path(key):
                error(
                    errors,
                    _(
                        "Step #%s (Logic): %s action #%s set.assign key '%s' must be a variable path "
                        "like 'serial_number' or 'http.status' (no 'vars.' prefix)."
                    )
                    % (idx, label, k, key),
                )
                continue
            _validate_expr(idx=idx, label=f'{label} action #{k} assign {key}', expr=raw_val, errors=errors)


def _validate_then(
    *,
    idx: int,
    label: str,
    then: Any,
    errors: list[str],
    step_id: str,
    step_ids: set[str],
    index_map: dict[str, int],
) -> None:
    if not isinstance(then, dict):
        error(errors, _('Step #%s (Logic): %s then is required and must be an object.') % (idx, label))
        return

    # Exactly one of: pass/goto/stop
    keys_present = 0
    if then.get('pass') is True:
        keys_present += 1
    if 'goto' in then:
        keys_present += 1
    if 'stop' in then:
        keys_present += 1

    if keys_present != 1:
        error(errors, _('Step #%s (Logic): %s then must specify exactly one of pass/goto/stop.') % (idx, label))
        return

    # pass
    if then.get('pass') is True:
        return

    # goto
    if 'goto' in then:
        tgt = str(then.get('goto') or '').strip()
        if not tgt:
            error(errors, _('Step #%s (Logic): %s goto must be a non-empty step id.') % (idx, label))
            return
        if tgt not in step_ids:
            error(errors, _("Step #%s (Logic): %s goto target '%s' does not match any step id.") % (idx, label, tgt))
            return

        cur_idx = index_map.get(step_id)
        tgt_idx = index_map.get(tgt)
        if cur_idx is None or tgt_idx is None:
            return
        if tgt_idx <= cur_idx:
            error(
                errors,
                _(
                    "Step #%s (Logic): %s goto target '%s' must refer to a later step (forward-only)."
                )
                % (idx, label, tgt),
            )
        return

    # stop
    stop_cfg = then.get('stop')
    if stop_cfg is True:
        return
    if isinstance(stop_cfg, dict):
        r = stop_cfg.get('reason')
        if r is not None and not isinstance(r, str):
            error(errors, _('Step #%s (Logic): %s stop.reason must be a string if provided.') % (idx, label))
        return

    error(errors, _('Step #%s (Logic): %s stop must be true or an object.') % (idx, label))


def _validate_expr(*, idx: int, label: str, expr: Any, errors: list[str]) -> None:
    """Validate expression shape only (Pass-1).

    Note:
        This does NOT evaluate. It validates that the expression tree is well-formed.
    """
    if expr is None:
        return

    if isinstance(expr, (str, int, float, bool, list)):
        return

    if not isinstance(expr, dict):
        error(errors, _('Step #%s (Logic): %s expression must be an object or primitive.') % (idx, label))
        return

    if 'path' in expr:
        p = expr.get('path')
        if not isinstance(p, str) or not p.strip():
            error(errors, _('Step #%s (Logic): %s path must be a non-empty string.') % (idx, label))
        return

    if 'const' in expr:
        return

    op = str(expr.get('op') or '').strip().lower()
    if op not in _ALLOWED_OPS:
        error(errors, _('Step #%s (Logic): %s has unknown op "%s".') % (idx, label, op))
        return

    if op in {'and', 'or'}:
        args = expr.get('args')
        if not isinstance(args, list) or not args:
            error(errors, _('Step #%s (Logic): %s %s requires a non-empty args array.') % (idx, label, op))
            return
        for i, a in enumerate(args, start=1):
            _validate_expr(idx=idx, label=f'{label}.{op}[{i}]', expr=a, errors=errors)
        return

    if op in {'not', 'exists', 'truthy', 'falsy'}:
        _validate_expr(idx=idx, label=f'{label}.{op}', expr=expr.get('arg'), errors=errors)
        return

    # binary ops
    _validate_expr(idx=idx, label=f'{label}.{op}.left', expr=expr.get('left'), errors=errors)
    _validate_expr(idx=idx, label=f'{label}.{op}.right', expr=expr.get('right'), errors=errors)

File: trustpoint/static/js/workflow/validators/logic.js
// trustpoint/static/js/workflow/validators/logic.js
// Client-side validation for Logic step params (mirrors server rules + blocks invalid numeric constants).

import { state } from '../state.js';

const SEGMENT_CHARS = new Set('abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_.'.split(''));

function isDotPath(s) {
  if (typeof s !== 'string' || !s) return false;
  for (const ch of s) if (!SEGMENT_CHARS.has(ch)) return false;

  const parts = s.split('.');
  return parts.every(
    (p) => p && /[A-Za-z_]/.test(p[0]) && /^[A-Za-z0-9]+$/.test(p.replaceAll('_', ''))
  );
}

function isBareVarPath(s) {
  if (!isDotPath(s)) return false;
  const t = String(s).trim();
  if (t.startsWith('vars.')) return false;
  if (t.startsWith('ctx.')) return false;
  return true;
}

function isPlainObject(v) {
  return (v && typeof v === 'object' && !Array.isArray(v));
}

const ALLOWED_OPS = new Set([
  'eq', 'ne', 'lt', 'lte', 'gt', 'gte',
  'and', 'or', 'not', 'exists', 'truthy', 'falsy',
  'add', 'sub', 'mul', 'div', 'mod',
]);

function push(errors, msg) {
  errors.push(msg);
}

function getStableStepIds() {
  const steps = Array.isArray(state.steps) ? state.steps : [];
  return steps.map((_, i) => `step-${i + 1}`);
}

function isNumericString(s) {
  if (typeof s !== 'string') return false;
  const t = s.trim();
  if (t === '') return false;
  const n = Number(t);
  return Number.isFinite(n);
}

function validateConstObject({ idx, label, obj, errors }) {
  // Supports UI form: { const: ..., _ui_type: ... }
  const uiType = String(obj._ui_type || '').trim().toLowerCase();

  if (!('const' in obj)) {
    push(errors, `Step #${idx} (Logic): ${label} const object is missing "const".`);
    return;
  }

  const v = obj.const;

  if (uiType === 'number') {
    // Strict rule to prevent silent corruption:
    // - Accept finite number OR numeric string.
    // - Reject everything else (including empty string).
    if (typeof v === 'number' && Number.isFinite(v)) return;
    if (typeof v === 'string' && isNumericString(v)) return;

    push(errors, `Step #${idx} (Logic): ${label} constant is marked as number but is not a valid number.`);
    return;
  }

  if (uiType === 'null') {
    if (v !== null) {
      push(errors, `Step #${idx} (Logic): ${label} constant is marked as null but is not null.`);
    }
    return;
  }

  if (uiType === 'boolean') {
    // Accept boolean; allow strings during typing but do not fail hard here unless you want strictness.
    // If you want strict: require typeof v === 'boolean'
    return;
  }

  // string or unknown: always acceptable
}

function validateThen({ idx, label, thenObj, errors, currentStableId }) {
  if (!thenObj || typeof thenObj !== 'object' || Array.isArray(thenObj)) {
    push(errors, `Step #${idx} (Logic): ${label} then is required and must be an object.`);
    return;
  }

  let keys = 0;
  if (thenObj.pass === true) keys += 1;
  if ('goto' in thenObj) keys += 1;
  if ('stop' in thenObj) keys += 1;

  if (keys !== 1) {
    push(errors, `Step #${idx} (Logic): ${label} then must specify exactly one of pass/goto/stop.`);
    return;
  }

  if (thenObj.pass === true) return;

  if ('goto' in thenObj) {
    const tgt = String(thenObj.goto || '').trim();
    if (!tgt) {
      push(errors, `Step #${idx} (Logic): ${label} goto must be a non-empty step id.`);
      return;
    }

    const ids = getStableStepIds();
    if (!ids.includes(tgt)) {
      push(errors, `Step #${idx} (Logic): ${label} goto target "${tgt}" does not match any step id.`);
      return;
    }

    const curIndex = ids.indexOf(currentStableId);
    const tgtIndex = ids.indexOf(tgt);
    if (curIndex >= 0 && tgtIndex >= 0 && tgtIndex <= curIndex) {
      push(errors, `Step #${idx} (Logic): ${label} goto target "${tgt}" must refer to a later step (forward-only).`);
    }
    return;
  }

  const stopCfg = thenObj.stop;
  if (stopCfg === true) return;
  if (stopCfg && typeof stopCfg === 'object' && !Array.isArray(stopCfg)) {
    if ('reason' in stopCfg && stopCfg.reason != null && typeof stopCfg.reason !== 'string') {
      push(errors, `Step #${idx} (Logic): ${label} stop.reason must be a string if provided.`);
    }
    return;
  }

  push(errors, `Step #${idx} (Logic): ${label} stop must be true or an object.`);
}

function validateExpr({ idx, label, expr, errors }) {
  if (expr == null) return;

  // primitives + lists are allowed as literals (matches server)
  const t = typeof expr;
  if (t === 'string' || t === 'number' || t === 'boolean') return;
  if (Array.isArray(expr)) return;

  if (!expr || typeof expr !== 'object') {
    push(errors, `Step #${idx} (Logic): ${label} expression must be an object or primitive.`);
    return;
  }

  // {"path": "..."}
  if ('path' in expr) {
    const p = expr.path;
    if (typeof p !== 'string' || !p.trim()) {
      push(errors, `Step #${idx} (Logic): ${label} path must be a non-empty string.`);
    }
    return;
  }

  // {"const": ...} (UI may include _ui_type metadata)
  if ('const' in expr) {
    if (isPlainObject(expr)) {
      validateConstObject({ idx, label, obj: expr, errors });
    }
    return;
  }

  const op = String(expr.op || '').trim().toLowerCase();
  if (!ALLOWED_OPS.has(op)) {
    push(errors, `Step #${idx} (Logic): ${label} has unknown op "${op}".`);
    return;
  }

  if (op === 'and' || op === 'or') {
    const args = expr.args;
    if (!Array.isArray(args) || args.length === 0) {
      push(errors, `Step #${idx} (Logic): ${label} ${op} requires a non-empty args array.`);
      return;
    }
    args.forEach((a, i) => {
      validateExpr({ idx, label: `${label}.${op}[${i + 1}]`, expr: a, errors });
    });
    return;
  }

  if (op === 'not' || op === 'exists' || op === 'truthy' || op === 'falsy') {
    validateExpr({ idx, label: `${label}.${op}`, expr: expr.arg, errors });
    return;
  }

  if (op === 'add' || op === 'sub' || op === 'mul' || op === 'div' || op === 'mod') {
    const args = expr.args;
    if (Array.isArray(args)) {
      if (args.length < 2) {
        push(errors, `Step #${idx} (Logic): ${label} ${op} requires args with at least 2 items.`);
        return;
      }
      args.forEach((a, i) => {
        validateExpr({ idx, label: `${label}.${op}[${i + 1}]`, expr: a, errors });
      });
      return;
    }

    // Defensive/legacy: allow left/right
    validateExpr({ idx, label: `${label}.${op}.left`, expr: expr.left, errors });
    validateExpr({ idx, label: `${label}.${op}.right`, expr: expr.right, errors });
    return;
  }

  // binary comparison ops: eq/ne/lt/lte/gt/gte
  if (!('left' in expr)) {
    push(errors, `Step #${idx} (Logic): ${label}.${op} is missing "left".`);
    return;
  }
  if (!('right' in expr)) {
    push(errors, `Step #${idx} (Logic): ${label}.${op} is missing "right".`);
    return;
  }

  validateExpr({ idx, label: `${label}.${op}.left`, expr: expr.left, errors });
  validateExpr({ idx, label: `${label}.${op}.right`, expr: expr.right, errors });
}

function validateActions({ idx, label, actions, errors }) {
  if (actions == null) return;
  if (!Array.isArray(actions)) {
    push(errors, `Step #${idx} (Logic): ${label} actions must be an array.`);
    return;
  }

  actions.forEach((act, k0) => {
    const k = k0 + 1;
    if (!act || typeof act !== 'object' || Array.isArray(act)) {
      push(errors, `Step #${idx} (Logic): ${label} action #${k} must be an object.`);
      return;
    }

    const type = String(act.type || '').trim().toLowerCase();
    if (type !== 'set') {
      push(errors, `Step #${idx} (Logic): ${label} action #${k} has unknown type "${type}".`);
      return;
    }

    const assign = act.assign;
    if (!assign || typeof assign !== 'object' || Array.isArray(assign)) {
      push(errors, `Step #${idx} (Logic): ${label} action #${k} set.assign must be an object.`);
      return;
    }

    Object.entries(assign).forEach(([rawKey, rawVal]) => {
      const key = String(rawKey || '').trim();
      if (!key || !isBareVarPath(key)) {
        push(
          errors,
          `Step #${idx} (Logic): ${label} action #${k} set.assign key "${key}" must be a variable path like "serial_number" or "http.status" (no "vars." prefix).`
        );
        return;
      }
      validateExpr({ idx, label: `${label} action #${k} assign ${key}`, expr: rawVal, errors });
    });
  });
}

function validateWhen({ idx, label, when, errors }) {
  if (Array.isArray(when)) {
    if (when.length === 0) {
      push(errors, `Step #${idx} (Logic): ${label} must not be an empty array.`);
      return;
    }
    when.forEach((e, i) => {
      validateExpr({ idx, label: `${label}[${i + 1}]`, expr: e, errors });
    });
    return;
  }

  validateExpr({ idx, label, expr: when, errors });
}

export function validateLogic(step, idx, errors) {
  const params = (step && step.params && typeof step.params === 'object' && !Array.isArray(step.params))
    ? step.params
    : {};

  const currentStableId = `step-${idx}`;

  const def = params.default;
  if (!def || typeof def !== 'object' || Array.isArray(def)) {
    push(errors, `Step #${idx} (Logic): default is required and must be an object.`);
  }

  let rules = params.rules;
  if (rules == null) rules = [];
  if (!Array.isArray(rules)) {
    push(errors, `Step #${idx} (Logic): rules must be an array if provided.`);
    return;
  }

  rules.forEach((rule, j0) => {
    const j = j0 + 1;

    if (!rule || typeof rule !== 'object' || Array.isArray(rule)) {
      push(errors, `Step #${idx} (Logic): rule #${j} must be an object.`);
      return;
    }

    const when = rule.when;
    if (when == null) {
      push(errors, `Step #${idx} (Logic): rule #${j} is missing "when".`);
    } else {
      validateWhen({ idx, label: `rule #${j} when`, when, errors });
    }

    validateActions({ idx, label: `rule #${j}`, actions: rule.actions, errors });
    validateThen({
      idx,
      label: `rule #${j}`,
      thenObj: rule.then,
      errors,
      currentStableId,
    });
  });

  if (def && typeof def === 'object' && !Array.isArray(def)) {
    validateActions({ idx, label: 'default', actions: def.actions, errors });
    validateThen({
      idx,
      label: 'default',
      thenObj: def.then,
      errors,
      currentStableId,
    });
  }
}
